### ISSUE TITLE
Key not found when calling world.Destroy

### ISSUE THREAD

**Issue by andreakarasho** [genaray/Arch#13]
```csharp

         struct T { };
         struct K { };

          var arch = new Type[] { typeof(T), typeof(K) };

            for (int i = 0; i< 100; ++i)
                _world.Create(arch);

            var rng = new Random();
            while (true)
            {
                _world.Query(new QueryDescription() { All = arch }, (in Entity e, ref T t, ref K k) =>
                {
                    if (rng.Next() % 2 == 0)
                    {
                        _world.Destroy(in e);
                    }
                });
            }
```

At the exception throwing the `_world.Size` is ` > 0`

**Comment by genaray** [comment]
Investigating :)

**Comment by genaray** [comment]
One thing that instantly comes into my mind is that you never should query and create/destroy entities at the same time.

Destroyed entities will be replaced with the last entity from the last used chunk, this is a mechanism used to always ensure dense packed arrays. 

When we query entities and remove them at the same time, the query will not be noticed about this. This may result in undefined behaviour, like that a lot of entities will not be queried since they were already moved to memory locations which the query already passed by. Or that the query probably runs to far since the underlaying chunks were modified. Its basically like you would do the following...

```csharp
foreach(var myItem in myList){ 
   myList.Remove(myItem); // Throws exception since we iterate and remove at the same time, a big no go even for standard lists
 }
```

## Future Solution

This behaviour will also NOT be altered, since such a notification mechanism would harm the query performance. However theres a feature planned called "CommandBuffer" which will allow you to record such entity modifications to push them to the world after the query finished.

## Current Solution

Add entities you wanna remove to a list, iterate over the list AFTER the query to destroy the entities.

**Comment by genaray** [comment]
> ```cs
>          struct T { };
>          struct K { };
> 
>           var arch = new Type[] { typeof(T), typeof(K) };
> 
>             for (int i = 0; i< 100; ++i)
>                 _world.Create(arch);
> 
>             var rng = new Random();
>             while (true)
>             {
>                 _world.Query(new QueryDescription() { All = arch }, (in Entity e, ref T t, ref K k) =>
>                 {
>                     if (rng.Next() % 2 == 0)
>                     {
>                         _world.Destroy(in e);
>                     }
>                 });
>             }
> ```
> 
> At the exception throwing the `_world.Size` is ` > 0`

A feature called `CommandBuffer` #17 is now in development and will make such behaviour possible :)

**Comment by genaray** [comment]
Solved with #24 :) which adds multiple command buffers to record and playback structural changes. 

```csharp
         struct T { };
         struct K { };

          var arch = new Type[] { typeof(T), typeof(K) };
          var destructionBuffer = new DestructionBuffer(world);          

            for (int i = 0; i< 100; ++i)
                _world.Create(arch);

            var rng = new Random();
            while (true)
            {
                _world.Query(new QueryDescription() { All = arch }, (in Entity e, ref T t, ref K k) =>
                {
                    if (rng.Next() % 2 == 0)
                    {
                       destructionBuffer.Destroy(in e);
                    }
                });
                destructionBuffer.Playback();
            }
```
